File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        answer =[]
        min_val=None
        max_val=None

        for interval in intervals:
            if(min_val is None):
                min_val=interval[0]
                max_val= interval[1]
                continue
            
            if(interval[0]<=max_val):
                min_val = min(min_val, interval[0])
                max_val = max(max_val, interval[1])
                continue
            if(interval[0]>max_val) :
                # no overlapping
                answer.append([min_val, max_val])
                min_val = interval[0]
                max_val = interval[1]
                continue
        
        answer.append([min_val, max_val])
      
        return answer

File: 0056-merge-intervals/0056-merge-intervals.ts
function merge(intervals: number[][]): number[][] {
    // merge all overlapping intervals,
    // return array of the non-overlapping intervals,
    // 주어진 구간들이 오름차순이 이미 되어있는가? x
    // intervals.length <= 10^4 이므로 O(n log n) 까지는 가능
    //[0,0] 잉 들어오면,
    // 1. sort the intervals by start time
    // 겹치는게 있으면 merge를 일단 먼저 한다.
    const results= []
    intervals.sort( (a,b) => a[0] - b[0] )
    let [prevStart, prevEnd] = intervals[0]
    for ( const [start, end] of intervals ) {
        // overlap인지 판단하기
        if( start <= prevEnd ) {
            // merge
            prevStart = Math.min(prevStart, start)
            prevEnd = Math.max(prevEnd, end)

        }else{
            //no overlap
            // push the previous interval to result
            results.push([prevStart, prevEnd].slice())
            prevStart = start
            prevEnd = end
        }

    }
    // 마지막 interval 추가
    results.push([prevStart, prevEnd].slice())
    return results
};
